src/yz_solrq.erl

```diff
@@ -0,0 +1,817 @@
+%%          flag to false (finally), and kick-starting the indexq.
+%% @end
+%%
+handle_cast(drain_complete, #state{indexqs = IndexQs} = State) ->
+    NewIndexQs = dict:fold(
+        fun(Index, #indexq{queue = Queue, queue_len = QueueLen, aux_queue = AuxQueue} = IndexQ, IndexQAccum) ->
+            NewIndexQ = IndexQ#indexq{
+                queue = queue:join(Queue, AuxQueue),
+                queue_len = QueueLen + queue:len(AuxQueue),
+                aux_queue = queue:new(),
+                draining = false
+            },
+            dict:store(Index, maybe_start_timer(Index, maybe_request_worker(Index, NewIndexQ)), IndexQAccum)
+        end,
+        dict:new(),
+        IndexQs
+    ),
+    {noreply, State#state{indexqs = NewIndexQs, drain_info = undefined}}.
+
+
+handle_info({flush, Index, HRef}, State) -> % timer has fired - request a worker
+    case find_indexq(Index, State) of
+        undefined ->
+            {noreply, State};
+        IndexQ ->
+            case IndexQ#indexq.href of
+                HRef ->
+                    {noreply, flush(Index, IndexQ#indexq{href = undefined}, State)};
+                _ -> % out of date
+                    {noreply, State}
+            end
+    end;
+handle_info(_Msg, State) ->
+    {noreply, State}.
+
+terminate(_Reason, _State) ->
+    ok.
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%%%===================================================================
+%%% Private
+%%%===================================================================
+
+
+%%
+%% @doc
+%% A drain is not in progress.  There are two cases to consider:
+%%   1. The result is ok; This means all batched messages were delivered; Reduce the
+%%      all_queue_len by the number of messages delivered.
+%%   2. The solrq_helper returned some undelivered messages; pre-pend these to the queue
+%%      for this index, and request a new worker, if we are over the requested minimum.
+%%      Reduce the all_queue_len by the number of messages delivered.
+%%
+%% Since the ACL has been adjusted, unblock any vnodes that might be waiting.
+%% @end
+%%
+handle_batch(
+    Index,
+    #indexq{draining = false, batch_start = T1} = IndexQ0,
+    Result,
+    State
+) ->
+    IndexQ1 =
+        case Result of
+            ok ->
+                yz_stat:batch_end(?YZ_TIME_ELAPSED(T1)),
+                IndexQ0#indexq{batch_start = undefined};
+            {retry, Undelivered} ->
+                requeue_undelivered(Undelivered, IndexQ0)
+        end,
+    IndexQ2 = maybe_request_worker(Index, IndexQ1),
+    IndexQ3 = maybe_start_timer(Index, IndexQ2),
+    update_indexq(Index, IndexQ3, State);
+
+%%
+%% @doc
+%% This queue is being drained.  There are two cases to consider:
+%%    1. The batch succeeded.  In this case, move any data we
+%%       have accumulated in aux_queue to the main queue,
+%%       and remove ourselves from the list
+%%       of remaining indices that need to be flushed.
```

Comment:
comments are out of date.  Need to review comments in this module, as there were several architectural changes between the time the comments were written and behavior was finalized.